### PR TITLE
Put in a deopt stub so we can observe guard failures.

### PR DESCRIPTION
**Pull request by ptersilie**
This allows us detect when a trace reaches deoptimisation (this strengthens the `simple.newcg.c` test again that we had to weaken in https://github.com/ykjit/yk/pull/1029#event-12104752848).

Requires rebasing once https://github.com/ykjit/yk/pull/1029#event-12104752848 is merged.

### CODE REVIEW

**Comment by vext01** [comment]
LGTM. Just one typo.

File: ykrt/src/compile/jitc_yk/codegen/x86_64.rs [hunk 1]
```diff
@@ -163,6 +168,12 @@ impl<'a> X64CodeGen<'a> {
         dynasm!(self.asm
             ; push rbp
             ; mov rbp, rsp
+            // Safe pointers to ctrlp_vars and frameaddr for later use.
```
[the remark this answers is not in the document]
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/codegen/x86_64.rs]: Fixed in[ed60324](https://github.com/ykjit/yk/pull/1030/commits/ed60324aaac7ea8db7063d60b63a83d760c0df0e)

**Comment by vext01** [comment]
Please squash

**Comment by ptersilie** [comment]
Squashed and rebased.

**Comment by vext01** [comment]
Retrying to see if it's transient.

**Comment by vext01** [comment]
This time git timed out.

**Comment by ltratt** [comment]
I saw a `git` process on b13 running for ages. Maybe `git gc` or similar?

**Comment by vext01** [comment]
It looks the the submodule clone timed out. We shouldn't see any `git gc` because we always clone afresh.

**Comment by ltratt** [comment]
We're not the only ones experiencing this https://github.com/rust-lang/rust/issues/122476

**Comment by ptersilie** [comment]
Another timeout...

**Comment by vext01** [comment]
The git timeouts should now be ~~fixed~~ worked-around for now, so now you are up against:
```
15:33:36 test lang_tests::simple.newcg.c ... FAILED
15:33:36 
15:33:36 failures:
15:33:36 
15:33:36 ---- lang_tests::simple.newcg.c status ----
15:33:36 Exited due to signal: 6
15:33:36 
15:33:36 ---- lang_tests::simple.newcg.c stderr ----
15:33:36 
15:33:36 Pattern (error at line 9):
15:33:36    ...
15:33:36    |...
15:33:36    |--- End aot ---
15:33:36    |...
15:33:36 >> |deopt...
15:33:36    |...
15:33:36 
15:33:36 Text (error at line 186):
15:33:36    ...
15:33:36    |note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
15:33:36    |thread 'simple.newcg' panicked at ykrt/src/mt.rs:576:51:
15:33:36    |called `Result::unwrap()` on an `Err` value: Any { .. }
15:33:36 >> |fatal runtime error: failed to initiate panic, error 5
15:33:36 
15:33:36 ---- lang_tests::simple.newcg.c stdout ----
```

Any ideas?

It's this line:
```
thread::spawn(|| do_compile()).join().unwrap();
```

which suggests the compilation thread crashed. If you add `RUST_BACKTRACE=full` you might get more info.

**Comment by ptersilie** [comment]
Running this manually on b13 also works. Giving `tryci` another go.

**Comment by ptersilie** [comment]
Can we try merging this again? I'm running out of ideas as to why this fails.

**Comment by vext01** [comment]
langtester swallowed the backtrace :P

**Comment by ltratt** [comment]
It didn't swallow the backtrace: the backtrace couldn't be generated (at least AFAICS).

FWIW I can replicate this with tryci on b16. If you scroll up you'll notice that the failure is caused by:

```
YKD_NEW_CODEGEN=1 RUST_TEST_SHUFFLE=1 cargo test --release
```

And with that I can also replicate things locally:

```
running 1 tests/vol/extra/ltratt/yk/tests/../bin/yk-config: 175: /vol/extra/ltratt/yk/target/release/ykllvm/bin/llvm-config: not found                                                  
thread '<unnamed>' panicked at tests/src/lib.rs:157:9:                          
yk-config exited with non-zero status                                           
stack backtrace:                        
   0:     0x556efe25d592 - std::backtrace_rs::backtrace::libunwind::trace::hfaaeee8ba270e11c
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x556efe25d592 - std::backtrace_rs::backtrace::trace_unsynchronized::h6fb1c4d325d0d969
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x556efe25d592 - std::sys_common::backtrace::_print_fmt::hd36d192d75c5bc13
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x556efe25d592 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hebc622152f918065
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x556efe284d1c - core::fmt::rt::Argument::fmt::h42d7ab9e6b9d76d9
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/core/src/fmt/rt.rs:142:9
   5:     0x556efe284d1c - core::fmt::write::hceb334e2bb0855cb                              
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/core/src/fmt/mod.rs:1153:17
   6:     0x556efe25a14f - std::io::Write::write_fmt::h21e1401d0661ae6d
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/io/mod.rs:1843:15
   7:     0x556efe25d364 - std::sys_common::backtrace::_print::hd5d84a1029af69b4
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x556efe25d364 - std::sys_common::backtrace::print::h6fd3aca52697573b
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x556efe25eb9b - std::panicking::default_hook::{{closure}}::h4a9c0b0282b315bd
  10:     0x556efe25e8f3 - std::panicking::default_hook::h64aacf0d5c35b498                  
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:292:9
  11:     0x556efe25f0ad - std::panicking::rust_panic_with_hook::h572774bd80531e18
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:779:13
  12:     0x556efe25ef49 - std::panicking::begin_panic_handler::{{closure}}::he9b751db0dede176
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:649:13
  13:     0x556efe25da66 - std::sys_common::backtrace::__rust_end_short_backtrace::h181bd77376b4f0f9
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/sys_common/backtrace.rs:171:18
  14:     0x556efe25ecb4 - rust_begin_unwind                                    
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:645:5
  15:     0x556efe283975 - core::panicking::panic_fmt::ha1f618a40e3acdf1       
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/core/src/panicking.rs:72:14
  16:     0x556efe12487c - tests::mk_compiler::h96ff993bcf18a620
                               at /vol/extra/ltratt/yk/tests/src/lib.rs:157:9   
  17:     0x556efe11f200 - c_tests::run_suite::{{closure}}::h31966eb9fad36ce9               
                               at /vol/extra/ltratt/yk/tests/langtest_c.rs:86:32
  18:     0x556efe1367dd - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h57316a3241e1d5b9
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/alloc/src/boxed.rs:2034:9
  19:     0x556efe1367dd - lang_tester::tester::run_tests::{{closure}}::h1f358c1135b1b4f5
                               at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lang_tester-0.8.0/src/tester.rs:745:43
  20:     0x556efe1367dd - std::panicking::try::do_call::h69f544ebadb288db                  
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:552:40
  21:     0x556efe1367dd - std::panicking::try::h55433574c60c8679
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:516:19
  22:     0x556efe1367dd - std::panic::catch_unwind::h9ba4c333c96a6662
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panic.rs:146:14
  23:     0x556efe1367dd - lang_tester::tester::run_tests::h7acc40be35393512
                               at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lang_tester-0.8.0/src/tester.rs:745:27
  24:     0x556efe130bf0 - lang_tester::tester::test_file::{{closure}}::hc6b187c4d6e2863b
                               at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lang_tester-0.8.0/src/tester.rs:679:24
  25:     0x556efe130bf0 - <F as threadpool::FnBox>::call_box::he28e6de6364dca3a
                               at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/threadpool-1.8.1/src/lib.rs:95:9
  26:     0x556efe159170 - threadpool::spawn_in_pool::{{closure}}::h65e38d2ca7ec7361
                               at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/threadpool-1.8.1/src/lib.rs:769:17
  27:     0x556efe159170 - std::sys_common::backtrace::__rust_begin_short_backtrace::hcb834e821de0acf5
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/sys_common/backtrace.rs:155:18
  28:     0x556efe15a491 - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::h33db9029dadfb944
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/thread/mod.rs:528:17
  29:     0x556efe15a491 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h70dcc4afd98a191c
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/core/src/panic/unwind_safe.rs:272:9
  30:     0x556efe15a491 - std::panicking::try::do_call::ha58028237ba13982
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:552:40
  31:     0x556efe15a491 - std::panicking::try::he0a024f1bc97d82e
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panicking.rs:516:19
  32:     0x556efe15a491 - std::panic::catch_unwind::h5500f69280136a6d
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/panic.rs:146:14
  33:     0x556efe15a491 - std::thread::Builder::spawn_unchecked_::{{closure}}::h980d92b0cd7bffdc
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/thread/mod.rs:527:30
  34:     0x556efe15a491 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hc2cee98d9472f161
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/core/src/ops/function.rs:250:5
  35:     0x556efe264c95 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h77b8b116ff3c4255
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/alloc/src/boxed.rs:2020:9
  36:     0x556efe264c95 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h994160effdff6027
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/alloc/src/boxed.rs:2020:9
  37:     0x556efe264c95 - std::sys::pal::unix::thread::Thread::new::thread_start::h336eeebfb56643a3
                               at /rustc/3cbb93223f33024db464a4df27a13c7cce870173/library/std/src/sys/pal/unix/thread.rs:108:17
  38:     0x7fbc50d3e134 - start_thread
                               at ./nptl/pthread_create.c:442:8
  39:     0x7fbc50dbe7dc - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  40:                0x0 - <unknown>
```

**Comment by ptersilie** [comment]
Note that the above appears to be an entirely different error though. CI was `thread 'ykrt/src/mt.rs:576:51:`, your output errors at ` tests/src/lib.rs:157:9`.

**Comment by ltratt** [comment]
The error in `mt` is a result of spawning a thread: it's telling us the other thread paniced and put the system in an unknown state.

**Comment by ptersilie** [comment]
Even with that command I still can't reproduce this on b16 with a fresh install.

**Comment by ltratt** [comment]
Please rebase against master and force push.

**Comment by ptersilie** [comment]
Rebased.

**Comment by ltratt** [comment]
OK so I'm gradually pinning the "real" error down:

```
thread '<unnamed>' panicked at /ci/.cargo/registry/src/index.crates.io-6f17d22bb
a15001f/typed-index-collections-3.1.0/src/slice/slice_index.rs:99:10:           
index out of bounds: the len is 1 but the index is 2                            
stack backtrace:                                                                
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose bac
ktrace.                                                                         
thread 'simple.newcg' panicked at ykrt/src/mt.rs:576:51:                        
called `Result::unwrap()` on an `Err` value: Any { .. }                         
stack backtrace:                                                                
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose bac
ktrace.                                                                         
fatal runtime error: failed to initiate panic, error 5    
```

Still not super helpful though...

**Comment by ltratt** [comment]
OK so I don't get a great backtrace but there is a tiny bit more info:

```
thread '<unnamed>' panicked at /ci/.cargo/registry/src/index.crates.io-6f17d22bb
a15001f/typed-index-collections-3.1.0/src/slice/slice_index.rs:99:10:           
index out of bounds: the len is 1 but the index is 2                            
stack backtrace:                                                                
   0:     0x7f82854d77c2 - <unknown>                                            
   1:     0x7f82854f856c - <unknown>                                            
   2:     0x7f82854d524f - <unknown>                                            
   3:     0x7f82854d7594 - <unknown>                                            
   4:     0x7f82854d8bbb - <unknown>                                            
   5:     0x7f82854d8913 - <unknown>                                            
   6:     0x7f82854d90cd - <unknown>                                            
   7:     0x7f82854d8fa2 - <unknown>                                            
   8:     0x7f82854d7c96 - <unknown>                                            
   9:     0x7f82854d8cd4 - <unknown>                                            
  10:     0x7f82854f7585 - <unknown>                                            
  11:     0x7f82854f77d2 - <unknown>                                            
  12:     0x7f8285491015 - <unknown>                                            
  13:     0x7f8285499b51 - <unknown>                                            
  14:     0x7f828548341a - <unknown>                                            
  15:     0x7f8285481b8b - <unknown>                                            
  16:     0x7f82854db0f5 - <unknown>                                            
  17:     0x7f82852ce134 - <unknown>                                            
  18:     0x7f828534da40 - clone                                                
  19:                0x0 - <unknown>                                            
thread 'simple.newcg' panicked at ykrt/src/mt.rs:576:51:                        
called `Result::unwrap()` on an `Err` value: Any { .. }                         
stack backtrace:                                                                
   0:     0x7f82854d77c2 - <unknown>                                            
   1:     0x7f82854f856c - <unknown>
   2:     0x7f82854d524f - <unknown>
   3:     0x7f82854d7594 - <unknown>
   4:     0x7f82854d8bbb - <unknown>
   5:     0x7f82854d8913 - <unknown>
   6:     0x7f82854d90cd - <unknown>
   7:     0x7f82854d8fa2 - <unknown>
   8:     0x7f82854d7c96 - <unknown>
   9:     0x7f82854d8cd4 - <unknown>
  10:     0x7f82854f7585 - <unknown>
  11:     0x7f82854f7a63 - <unknown>
  12:     0x7f8285494e52 - <unknown>
  13:     0x7f8285494496 - <unknown>
  14:     0x7f82854686ad - __ykrt_control_point
  15:           0x202ccf - main
  16:     0x7f828526c24a - <unknown>
  17:     0x7f828526c305 - __libc_start_main
  18:           0x202a71 - _start
  19:                0x0 - <unknown>
fatal runtime error: failed to initiate panic, error 5
```

**Comment by vext01** [comment]
`RUST_BACKTRACE=full`?

**Comment by ltratt** [comment]
That's with `RUST_BACKTRACE=full`. My guess is the "unknown" stuff is perhaps LLVM calls that don't have debug info? I dunno though.

**Comment by vext01** [comment]
> LLVM calls that don't have debug info

Plausible. LLVM debug info is enormous, so we don't enable it.

**Comment by vext01** [comment]
Where do we stand with this one?

**Comment by ltratt** [comment]
Please force push a rebase against master.

**Comment by ptersilie** [comment]
Rebased. Note the new commit [f04b56f](https://github.com/ykjit/yk/pull/1030/commits/f04b56fa2042b5045d56569fca530503fee7c4d3) which enables us to actually test that deopt is "working".

**Pull request merged**